lib/src/layer/tile_layer/tile_provider/network/caching/tile_metadata.dart

```diff
@@ -0,0 +1,79 @@
+import 'dart:io' show HttpDate; // this is web safe!
+
+import 'package:flutter_map/flutter_map.dart';
+import 'package:meta/meta.dart';
+
+/// Metadata about a tile cached with a [MapCachingProvider]
+///
+/// Map caching is based on HTTP headers, which this class contains and can
+/// encode/decode to/from JSON.
+///
+/// External usage of this class is not usually necessary. It is visible so
+/// other tile providers may make use of it.
+@immutable
+class CachedMapTileMetadata {
+  /// Create new metadata
+  ///
+  /// [lastModifiedLocally] must be set to [DateTime.timestamp]. Other
+  /// properties should usually be set based on the tile's HTTP response
+  /// headers.
+  const CachedMapTileMetadata({
+    required this.lastModifiedLocally,
+    required this.staleAt,
+    required this.lastModified,
+    required this.etag,
+  });
+
+  /// Decode metadata from JSON
+  CachedMapTileMetadata.fromJson(Map<String, dynamic> json)
+      : lastModifiedLocally =
+            HttpDate.parse(json['lastModifiedLocally'] as String),
+        staleAt = HttpDate.parse(json['staleAt'] as String),
+        lastModified = json.containsKey('lastModified') &&
+                (json['lastModified'] as String).isNotEmpty
+            ? HttpDate.parse(json['lastModified'] as String)
+            : null,
+        etag = json.containsKey('etag') && (json['etag'] as String).isNotEmpty
+            ? json['etag'] as String
+            : null;
+
+  /// Used to efficiently allow updates to already cached tiles
+  ///
+  /// Must be set to [DateTime.timestamp] when a new tile is cached or a tile
+  /// is updated.
+  final DateTime lastModifiedLocally;
+
+  /// The date/time at which the tile becomes stale according to the HTTP spec
+  final DateTime staleAt;
+
+  /// The tile's last modified HTTP header
+  final DateTime? lastModified;
+
+  /// The tile's etag HTTP header
+  final String? etag;
+
+  /// Whether the tile is currently stale
+  bool get isStale => DateTime.timestamp().isAfter(staleAt);
+
+  /// Encode the metadata to JSON
+  Map<String, String> toJson() => {
+        'lastModifiedLocally': HttpDate.format(lastModifiedLocally),
+        'staleAt': HttpDate.format(staleAt),
+        if (lastModified != null)
+          'lastModified': HttpDate.format(lastModified!),
+        if (etag != null) 'etag': etag!,
+      };
+
+  @override
+  int get hashCode =>
+      Object.hash(lastModifiedLocally, staleAt, lastModified, etag);
+
+  @override
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      (other is CachedMapTileMetadata &&
+          lastModifiedLocally == other.lastModifiedLocally &&
+          staleAt == other.staleAt &&
+          lastModified == other.lastModified &&
+          etag == other.etag);
```

Comment:
I personally find the [equatable](https://pub.dev/packages/equatable) package to handle these == and hashCode operators really good. This is just an idea in case you haven't heard about it, but it would essentially replace your hashCode and == functions with this:

```dart
class CachedMapTileMetadata extends Equatable {
//...
  @override
  List<Object?> get props => <dynamic>[
        lastModifiedLocally,
        staleAt,
        lastModified,
        etag,
      ];
```

This is just an idea in case it helps, feel free to resolve this comment

Reply:
It might make sense to use equatable, but probably best to introduce it all at once in a separate PR. I've avoided it before just because I don't want to take the time to set it up everywhere (even though it's faster in the long term). I've looked for a good IDE plugin in the past, but I haven't found a reliable one.